File: app/src/main/java/com/supercilex/robotscouter/ui/scout/ScoutValueDialogBase.java
package com.supercilex.robotscouter.ui.scout;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.design.widget.TextInputLayout;
import android.view.View;
import android.widget.EditText;

import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.RobotScouter;
import com.supercilex.robotscouter.ui.KeyboardDialogBase;
import com.supercilex.robotscouter.util.DatabaseHelper;

public abstract class ScoutValueDialogBase<T> extends KeyboardDialogBase implements Runnable {
    protected static final String CURRENT_VALUE = "current_value";

    protected TextInputLayout mInputLayout;
    protected EditText mValue;

    protected abstract T getValue();

    @StringRes
    protected abstract int getTitle();

    @StringRes
    protected abstract int getHint();

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        View rootView = View.inflate(getContext(), R.layout.dialog_scout_value, null);
        mInputLayout =
                (TextInputLayout) rootView.findViewById(R.id.value_layout);
        mInputLayout.setHint(getString(getHint()));
        mValue = (EditText) mInputLayout.findViewById(R.id.value);
        mValue.setText(getArguments().getString(CURRENT_VALUE));
        mValue.post(this);

        return createDialog(rootView, getTitle());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        RobotScouter.getRefWatcher(getActivity()).watch(this);
    }

    @Override
    public void run() {
        mValue.selectAll();
    }

    @Override
    protected EditText getLastEditText() {
        return mValue;
    }

    @Override
    protected boolean onClick() {
        DatabaseHelper.getRef(getArguments()).setValue(getValue());
        return true;
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scout/ScoutValueDialogBase.kt
package com.supercilex.robotscouter.ui.scout

import android.app.Dialog
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.design.widget.TextInputLayout
import android.view.View
import android.widget.EditText

import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.RobotScouter
import com.supercilex.robotscouter.ui.KeyboardDialogBase
import com.supercilex.robotscouter.util.DatabaseHelper

abstract class ScoutValueDialogBase<out T> : KeyboardDialogBase() {
    private val rootView: View by lazy { View.inflate(context, R.layout.dialog_scout_value, null) }
    protected val inputLayout: TextInputLayout by lazy { rootView.findViewById<TextInputLayout>(R.id.value_layout) }
    override val lastEditText: EditText by lazy { inputLayout.findViewById<EditText>(R.id.value) }

    protected abstract val value: T?
    @get:StringRes protected abstract val title: Int
    @get:StringRes protected abstract val hint: Int

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        inputLayout.hint = getString(hint)
        lastEditText.apply {
            setText(arguments.getString(CURRENT_VALUE))
            post { selectAll() }
        }

        return createDialog(rootView, title)
    }

    override fun onDestroy() {
        super.onDestroy()
        RobotScouter.getRefWatcher(activity).watch(this)
    }

    override fun onClick(): Boolean {
        DatabaseHelper.getRef(arguments).setValue(value)
        return true
    }

    protected companion object {
        @JvmStatic val CURRENT_VALUE = "current_value"
    }
}
